Image-to-Image Task Guide (#26595)
huggingface/transformers: docs/source/en/tasks/image_to_image.md

<!--Copyright 2023 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
rendered properly in your Markdown viewer.

-->

# Image-to-Image Task Guide

> 

Reviewer: ```suggestion

[[open-in-colab]]

```

Reviewer: So that folks can open the doc in Colab :)

> Image-to-Image task is the task where an application receives an image and outputs another image. This has various subtasks, including, image enhancement (super resolution, low light enhancement, deraining and so on), image inpainting, and more. 

Reviewer 2: ```suggestion
Image-to-Image task is the task where an application receives an image and outputs another image. This has various subtasks, including image enhancement (super resolution, low light enhancement, deraining and so on), image inpainting, and more. 
```

This guide will show you how to:

> - Use an image-to-image pipeline for super resolution task,

Reviewer 3: :heavy_plus_sign: I also like this idea

> - Use an image-to-image pipeline for super resolution task,

Author: IMHO they're both very narrow topics to actually make a header.

> - Use an image-to-image pipeline for super resolution task,

Reviewer 3: Alright! Totally fine for me. :)

- Run image-to-image models for same task without a pipeline.

Note that as of the time this guide is released, `image-to-image` pipeline only supports super resolution task.

Let's begin by installing the necessary libraries.

```bash
pip install transformers
```

> We can now initialize the pipeline with a [Swin2SR model](https://huggingface.co/caidas/swin2SR-lightweight-x2-64). We can then infer with the pipeline by calling it with an image.

Reviewer: Is this the only model supported by this pipeline? If there are others, it would be cool to mention them.

Author: Swin2SR is the only model class that is supported, hence I wrote a Swin2SR model, but other checkpoints can be used. Should I still clarify=

Reviewer: Usually, we have pipelines that support more than one model, so maybe it's worth adding one sentence saying that currently, this is the only supported model for this pipeline.

Author: I addressed this comment, thank you!

```python
from transformers import pipeline

> image = Image.open("/content/cat.png")

Reviewer: It looks like you don't need this line (you're loading the image from a URL just below)

Author: ah yes, it was left from my notebook apparently. Fixed this.

pipe = pipeline(task="image-to-image", model="caidas/swin2SR-lightweight-x2-64", device=0)
```

Now, let's load an image.

```python
from PIL import Image
import requests

url = "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/tasks/cat.jpg"
image = Image.open(requests.get(url, stream=True).raw)

print(image.size)

> # (532, 432)

Reviewer: For the printed outputs, you can do either 
1. 
\```bash
(532, 432)
\```
or, format code example like this:
\```python
\>>> from PIL import Image
\>>> import requests

\>>> url = "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/tasks/cat.jpg"
\>>> image = Image.open(requests.get(url, stream=True).raw)

\>>> print(image.size)
(532, 432)
\```

```

> ![Image](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/tasks/cat.jpg)

Reviewer: ```suggestion
<div class="flex justify-center">
     <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/tasks/cat.jpg" alt="Photo of a cat"/>
</div>
```

> We can now do inference with pipeline. We will get an upscaled version of the cat image. 

Reviewer 2: ```suggestion
We can now do inference with the pipeline. We will get an upscaled version of the cat image. 
```

> ```python

Reviewer: Same thing for the output. Please choose one of the two existing formatting options.

upscaled = pipe(image)
print(upscaled.size)

# (1072, 880)
```

If you wish to do inference yourself with no pipeline, you can use the `Swin2SRForImageSuperResolution` and `Swin2SRImageProcessor` classes of transformers. We will use the same model checkpoint for this. Let's initialize the model and the processor.

```python
from transformers import Swin2SRForImageSuperResolution, Swin2SRImageProcessor 

> model = Swin2SRForImageSuperResolution.from_pretrained("caidas/swin2SR-lightweight-x2-64").to(device=0)

Reviewer 3: Not sure if this model is also suited for CPU.
But maybe having
```python
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
```
for all parts where `device=0` is used.

Author: Addressed this.

processor = Swin2SRImageProcessor("caidas/swin2SR-lightweight-x2-64")
```

> `pipeline` abstracts away the preprocessing and postprocessing steps that we have to do ourselves, so let's preprocess the image. We will pass the image to processor and then take the pixel values to GPU. 

Reviewer 2: I'd recommend using torch.device = "cuda:0" if torch.cuda.is_available() else "cpu" as some people will run this on a CPU

```python
pixel_values = processor(image, return_tensors="pt").pixel_values
print(pixel_values.shape)

pixel_values = pixel_values.to(device=0)
```

We can now infer the image by passing pixel values to the model.

```python
import torch

with torch.no_grad():
  outputs = model(pixel_values)
```
Output is an object of type `ImageSuperResolutionOutput` that looks like below 👇 

```
(loss=None, reconstruction=tensor([[[[0.8270, 0.8269, 0.8275,  ..., 0.7463, 0.7446, 0.7453],
          [0.8287, 0.8278, 0.8283,  ..., 0.7451, 0.7448, 0.7457],
          [0.8280, 0.8273, 0.8269,  ..., 0.7447, 0.7446, 0.7452],
          ...,
          [0.5923, 0.5933, 0.5924,  ..., 0.0697, 0.0695, 0.0706],
          [0.5926, 0.5932, 0.5926,  ..., 0.0673, 0.0687, 0.0705],
          [0.5927, 0.5914, 0.5922,  ..., 0.0664, 0.0694, 0.0718]]]],
       device='cuda:0'), hidden_states=None, attentions=None)
```
We need to get the `reconstruction` and post-process it for visualization. Let's see how it looks like.

```python
outputs.reconstruction.data.shape
# torch.Size([1, 3, 880, 1072])
```

> We need to squeeze the output and get rid of axis 0, clip the values, then convert it to be numpy float. Then we will arrange axes to have the shape [1072, 880], and finally, bring the output back to range [0, 255].

Reviewer 2: Squeezing and getting rid of axis 0 is the same :)

Author: yes, I made it as an explanation (i.e. when you call squeeze you get rid of axis 0)

```python
import numpy as np

# squeeze, take to CPU and clip the values
output = outputs.reconstruction.data.squeeze().cpu().clamp_(0, 1).numpy()
# rearrange the axes
output = np.moveaxis(output, source=0, destination=-1)
# bring values back to pixel values range
output = (output * 255.0).round().astype(np.uint8)
Image.fromarray(output)
```

> ![Image](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/tasks/cat_upscaled.png)

Reviewer: ```suggestion
<div class="flex justify-center">
     <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/tasks/cat_upscaled.png" alt="Upscaled photo of a cat"/>
</div>
```
